fe/fe-core/src/main/java/com/starrocks/connector/AsyncTaskQueue.java

```diff
@@ -0,0 +1,265 @@
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Executor;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * AsyncTaskQueue is a class that manages and executes asynchronous tasks
+ * <p>
+ * The class uses an Executor to run tasks concurrently. It maintains two queues:
+ * - outputQueue: stores the results of completed tasks.
+ * - taskQueue: stores the tasks that need to be executed.
+ * <p>
+ * The class also keeps track of the number of currently running tasks using an AtomicInteger.
+ * It provides methods to start tasks, retrieve outputs, and handle task exceptions.
+ * <p>
+ * Make sure there is a single consumer.
+ */
+
+/*
+ * Here is a code example:
+ * AsyncTaskQueue<String> asyncTaskQueue = new AsyncTaskQueue<>(executorService);
+ * asyncTaskQueue.setMaxRunningTaskCount(maxRunningTaskCount);
+ * asyncTaskQueue.setMaxOutputQueueSize(maxOutputQueueSize);
+ * asyncTaskQueue.start(tasks);
+ * while(asyncTaskQueue.hasMoreOutput()) {
+ *     List values = asyncTaskQueue.getOutputs();
+ * }
+ */
+
+public class AsyncTaskQueue<T> {
+
+    public interface Task<T> {
+        List<T> run() throws InterruptedException;
+
+        default List<Task<T>> moreTasks() {
+            return null;
+        }
+
+        default boolean isDone() {
+            return true;
+        }
+    }
+
+    public AsyncTaskQueue(Executor executor) {
+        this.executor = executor;
+    }
+
+    // output queue
+    ReentrantLock outputQueueLock = new ReentrantLock();
+    Condition outputQueueCondition = outputQueueLock.newCondition();
+    ArrayList<T> outputQueue = new ArrayList<>();
+    AtomicInteger outputQueueSize = new AtomicInteger();
+    boolean hasMoreOutput = true;
+
+    // task queue
+    ConcurrentLinkedDeque<Task<T>> taskQueue = new ConcurrentLinkedDeque<>();
+    AtomicInteger runningTaskCount = new AtomicInteger(0);
+    AtomicInteger taskQueueSize = new AtomicInteger(0);
+    AtomicReference<Exception> taskException = new AtomicReference<>(null);
+
+    // ---------------
+    int maxRunningTaskCount = Integer.MAX_VALUE;
+    int maxOutputQueueSize = Integer.MAX_VALUE;
+    Executor executor;
+
+    public void setMaxRunningTaskCount(int maxRunningTaskCount) {
+        this.maxRunningTaskCount = maxRunningTaskCount;
+    }
+
+    public void setMaxOutputQueueSize(int maxOutputQueueSize) {
+        this.maxOutputQueueSize = maxOutputQueueSize;
+    }
+
+    public void start(List<? extends Task<T>> tasks) {
+        taskQueue.addAll(tasks);
+        taskQueueSize.addAndGet(tasks.size());
+        // or we have to trigger tasks?
+        triggerTask();
+    }
+
+    public int computeOutputSize(T output) {
+        return 1;
+    }
+
+    private void tryGetOutputs(List<T> outputs, int maxSize) {
```

Comment:
can we use future<T>? we can use future.get() to get the outputs

Reply:
this function is private, no need to make it as future.

about the `getOutputs()`, we can make it as a future.  But I don't see any benefit right now.